[ESP32] drivers: Add UART Driver with FIFO/Interrupt support

ExtremeGTX opened the pull request:
- Fixes #3981
- Implement UART Polling functions
- Implement UART Interrupt APIs
- Remove dependency on esp32_rom_uart_xxx functions
- Update Device tree with UART addresses and pin config
- Update ESP32 UART KConfig

### Notes about implementation:
- Interrupts now defined as a local macros, and should be removed
later on, when interrupts for esp32 are supported in dts

- Threshold interrupts are used for TX/RX
- Reseting FIFOs using _RST bit will corrupt FIFO of UART2 when used for
UART1 and vice-versa, so a generic way is used for all three UARTs

- Old Silicon rev is not supported


### Related issues:
- #3471 - Espressif ESP Architecture Support

### Tests:
 - uart_async_test
 - Polling functions
 - Different baud rates
 - multiple UART simultaneous operations
 - Console Subsystem
 - Shell Subsystem
 
 
### HW Features not implemented:
- [ ] RS485 Mode
- [ ] IRDA Mode
- [ ] Baud Rate Auto Detection

### Zephyr Features not implemented:
- [ ] Async IO APIs
- [ ] Power Management
- [ ] Adapt to System clock frequency changes, it uses the `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC`

### Questions:
- `uart_configure` vs `line_ctrl` do we need both ?
- One weird thing still needs some investigation is `uart_async_test` success even if Console I/O fails in a real application.
- setting `zephyr,shell-uart=&uart1` in dts and setting `(UART_1) Device Name of UART Device for SHELL_BACKEND_SERIAL` ? KConfig solution work, dts has no effect
- I think we need to have `esp32_disable_peripheral` function in soc.h
- There is no checking if `CONFIG_UART_CONSOLE_ON_DEV_NAME` or `SHELL_BACKEND_SERIAL` has a valid device name, the device will boot to crash 

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

ExtremeGTX committed:
drivers: esp32: Add UART Driver with FIFO/Interrupt support

- Fixes #3981
- Implement UART Polling functions
- Implement UART Interrupt APIs
- Remove dependency on esp32_rom_uart_xxx functions
- Update Device tree with UART addresses and pin config
- Update ESP32 UART KConfig

Notes about implementation:
- Interrupts now defined as a local macros, and should be removed
later on, when interrupts for esp32 are supported in dts

- Threshold interrupts are used for TX/RX
- Reseting FIFOs using _RST bit will corrupt FIFO of UART2 when used for
UART1 and vice-versa, so a generic way is used for all three UARTs

- Old Silicon rev is not supported

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

dcpleung commented on dts/xtensa/espressif/esp32.dtsi:
```diff
@@ -31,11 +31,27 @@
-			reg = <0x40008fd0 0x400>;
-			label = "ROMUART";
+		uart0: uart@3ff40000 {
+			compatible = "xtensa,esp32-uart";
```
I think we should keep "espressif" as vendor. "xtensa" is not exactly a vendor.

galak replied:
> I think we should keep "espressif" as vendor. "xtensa" is not exactly a vendor.

Is the IP for the UART coming from 'xtensa' or 'espressif'?

ExtremeGTX replied:
@galak @dcpleung 

I think it is not easy to be 100% sure if IP comes with Xtensa or made by espressif, i tried to do a quick search but there is no solid info about it AFAIK.

i think the fix would be to rename file `xtensa,esp32-uart.yaml` to `espressif,esp32-uart.yaml` because `xtensa` is not a vendor and also to keep naming consistent in `dts/binding/`.
then update the `compatible` value in the dts.

feel free to suggest alternatives

nashif replied:
the IP is owned by espressif. Using xtensa here is wrong if we are looking for a vendor, xtensa is not a vendor (Cadence is)

ExtremeGTX replied:
so, the correct naming is `cadence,esp32-uart` or `cadence,espressif-esp32-uart` ?

nashif replied:
I think it should be espressif. They own the IP, not cadence.

ydamigos replied:
I agree with @nashif. It should be `espressif,esp32-uart` and `espressif,esp32-uart.yaml`. Cadence owns Tensilica Xtensa processors, such as the Xtensa LX6 dataplane processing units (DPUs).

ydamigos commented on drivers/serial/uart_esp32.c:
```diff
@@ -1,58 +1,541 @@
+	.pins = {								\
+		.tx = DT_INST_##idx##_XTENSA_ESP32_UART_TX_PIN,			\
+		.rx = DT_INST_##idx##_XTENSA_ESP32_UART_RX_PIN,			\
+		.rts = DT_INST_##idx##_XTENSA_ESP32_UART_RTS_PIN,		\
```
Setting RTS and CTS should be optional as it is in DT. If they are not defined in boards DT, building uart driver fails.
```
[ 91%] Building C object zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_esp32.c.obj
zephyr/drivers/serial/uart_esp32.c:500:10: error: 'DT_INST_0_XTENSA_ESP32_UART_RTS_PIN' undeclared here (not in a function); did you mean 'DT_INST_0_XTENSA_ESP32_UART_RX_PIN'?
   .rts = DT_INST_##idx##_XTENSA_ESP32_UART_RTS_PIN,  \
          ^~~~~~~~
zephyr/drivers/serial/uart_esp32.c:532:1: note: in expansion of macro 'ESP32_UART_INIT'
 ESP32_UART_INIT(0);
 ^~~~~~~~~~~~~~~
zephyr/drivers/serial/uart_esp32.c:501:10: error: 'DT_INST_0_XTENSA_ESP32_UART_CTS_PIN' undeclared here (not in a function); did you mean 'DT_INST_0_XTENSA_ESP32_UART_TX_PIN'?
   .cts = DT_INST_##idx##_XTENSA_ESP32_UART_CTS_PIN,  \
          ^~~~~~~~
zephyr/drivers/serial/uart_esp32.c:532:1: note: in expansion of macro 'ESP32_UART_INIT'
 ESP32_UART_INIT(0);
 ^~~~~~~~~~~~~~~
make[2]: *** [zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/build.make:63: zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_esp32.c.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:1497: zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

ExtremeGTX replied:
Oh, I forgot about that !
Nice catch.
I will add also UART_X_FLOW_CONTROL to Kconfig to control the feature, dunno if there is a way through dts

ydamigos replied:
Linux uses `uart-has-rtscts` optional property to indicate RTS/CTS hardware flow control and the optional properties  rts-gpios, cts-gpios for RTS/CTS pins. You could check Linux's [Generic Serial DT Bindings](https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/serial/serial.txt). 

Better to add them in `uart.yaml` because most of UART IP blocks support RTS/CTS flow control.

ExtremeGTX replied:
@ydamigos i have 2 questions:
1. should i move also tx,rx,cts,rts-pin to uart.yaml ?
2. should i rename `cts/rts-pin` to be like linux `cts/rts-gpios` ?

ydamigos replied:
> should i move also tx,rx,cts,rts-pin to uart.yaml ?

I would suggest only cts and rts.

> should i rename cts/rts-pin to be like linux cts/rts-gpios ?

Yes, we try to keep zephyr's DT as close as possible to linux's DT.

nashif commented:
@ExtremeGTX which samples/tests did you test this with? I just tried the shell, and still does not work for me.

ExtremeGTX commented:
@nashif i enabled the console in kconfig and used `samples/subsys/console/echo/src/main.c` to test if i could send and receive a char.

also, i tried with Shell enabled in KConfig and i can get the basic shell working 
FYI, if both console and shell are enabled on UART_0 only console will work.

To test them simultaneously, I have set UART_0 for Shell and UART_2 for Console. 
Remember, old silicon rev. will not work, because of the problem in writing to FIFO Addresses (It's fixed in Silicon Rev. 1).
https://elinux.org/images/c/cd/Pincontrol-gpio-update.pdf (Section 3.3)

nashif commented:
>     * Old Silicon rev is not supported

how do I know if my rev is supported?

ExtremeGTX commented:
Here is the output of the esptool.py during flashing, you can see chip version on line:4
Unfortunately, there is no sign on the chip packaging it self to indicate the version.

```
esptool.py v2.7-dev
Serial port /dev/ttyS0
Connecting........__
Chip is ESP32D0WDQ6 (revision 1)
Features: WiFi, BT, Dual Core, Coding Scheme None
MAC: 30:ae:a4:4f:a3:94
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Auto-detected Flash size: 4MB
Flash params set to 0x0220
Wrote 49152 bytes at 0x00001000 in 4.4 seconds (88.9 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
```

ExtremeGTX committed:
drivers: esp32: UART Fix yaml/dts binding name

replace "xtensa" with "espressif", as the latter is the correct
vendor name.

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

ExtremeGTX committed:
dts: uart: move hw-flow-control to uart.yaml

- hw-flow-control added to uart.yaml as it is a common feature
- cleanup other related yaml files
- change property 'category' to 'rquired' in yaml files

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

ExtremeGTX committed:
drivers: esp32: UART - use dts config for HW flow Control

- dts updated for UART1/2
- Additional check added to _INIT macro to configure flow control mode
- Additional check added to _INIT macro to set CTS/RTS gpios values
- Additional check added for gpio config

Signed-off-by: Mohamed ElShahawi <ExtremeGTX@hotmail.com>

ydamigos reviewed (changes requested):
Looks good. Tested it both console and shell on odroid-go. They worked fine.
Just a minor comment. Could you squash commits  de94bac158e66f16c604ebefccbde4fea9942a18 and 47b5c8ea3c66c847b0d0f828224c25c3c0f73ab6 in commit 80c25c3023a38be703f416f04039b1ea35e07b83.

Please also rebase to resolve conflicts.

I tried also sample shell_module. It doesn't work. It seems that it doesn't boot. It is hard to debug it on odroid-go as the debug pins aren't available.

ExtremeGTX commented:
@ydamigos 
squashing and rebasing done.

about the crash you face, i think there is a problem with arch support (still not sure), because i tried to debug logger subsys and i got `LoadStoreError` then `IllegalInstruction`.
tried to ask for support on slack but i think the Guys are very busy.

I have a JTAG setup so feel free to suggest a tips/workarounds/docs etc.

ydamigos commented on drivers/serial/uart_esp32.c:
```diff
@@ -472,59 +476,66 @@ static const struct uart_driver_api uart_esp32_api = {
-static const struct uart_esp32_config uart_esp32_cfg_port_##idx = {		\
-	.dev_conf = {								\
-		.base = (u8_t *)DT_INST_##idx##_ESPRESSIF_ESP32_UART_BASE_ADDRESS,	\
+#define ESP32_UART_INIT(idx)									      \
```
Please align `\`.

ydamigos reviewed (approved):
Looks good. Just a small alignment in macros.

dcpleung commented on dts/xtensa/espressif/esp32.dtsi:
```diff
@@ -31,11 +31,27 @@
-			reg = <0x40008fd0 0x400>;
-			label = "ROMUART";
+			reg = <0x3ff40000 0x400>;
+			//interrupts = <12>;
```
Is commenting out the `interrupts` intentional? If this is intentional, maybe add a comment on why.

Also, our coding guideline says to use `/* */` instead of `//` for comments.

ExtremeGTX replied:
> Is commenting out the `interrupts` intentional? If this is intentional, maybe add a comment on why.

yes, because esp32 interrupt controller still not supported in device tree.
as a temp solution there is 3 macros in `uart_esp32.c` for IRQ numbers.

> Also, our coding guideline says to use `/* */` instead of `//` for comments.

dunno what is the best option here:
1. implement a dummy interrupt-controller node for esp32 then remove `//`
2. change this line to 
```
/* interrupts = <12>; - TODO: Enable interrupts when interrupt-controller got supported in device tree*/
```
3. remove this line without leaving any notes/comments

ExtremeGTX replied:
One more thing
do we need a copyright notice for `dts/bindings/serial/espressif,esp32-uart.yaml` ?
if yes should it be me or the one who created the file ?

dcpleung replied:
TODO (or better FIXME) would be fine. It acts as a reminder that we need to do something about it.

Not sure about the copyright. Since we have other yaml files without copyright, maybe just leave it as-is, for now? We can always go back with git log and figure it out later.

erwango reviewed (commented):
Would be nice to clean the first commit message.
Otherwise LGTM
